buildSrc/src/main/java/com/google/firebase/gradle/ReleaseGenerator.kt

```diff
@@ -0,0 +1,118 @@
+package com.google.firebase.gradle
+
+import java.io.File
+import java.io.IOException
+import org.eclipse.jgit.api.Git
+import org.eclipse.jgit.api.ListBranchCommand
+import org.eclipse.jgit.api.errors.GitAPIException
+import org.eclipse.jgit.lib.Constants
+import org.eclipse.jgit.lib.ObjectId
+import org.eclipse.jgit.lib.Ref
+import org.gradle.api.DefaultTask
+import org.gradle.api.tasks.TaskAction
+
+open class ReleaseGenerator : DefaultTask() {
+    @TaskAction
+    @Throws(Exception::class)
+    fun generateReleaseConfig() {
+        val currentRelease = project.property("currentRelease").toString()
+        val pastRelease = project.property("pastRelease").toString()
+        val rootDir = project.rootDir
+        val availableModules = parseSubProjects(rootDir)
+
+        val repo = initRepo(rootDir)
+        val headRef = repo.repository.resolve(Constants.HEAD)
+        val branchRef = getObjectRefForBranchName(repo, pastRelease)
+
+        val changedDirs = getChangedLibraries(repo, branchRef, headRef, FIREBASE_LIBRARIES)
+        val changedModules = findChangedModules(changedDirs, availableModules)
+        writeReleaseConfig(rootDir, changedModules, currentRelease)
+    }
+
+    private fun findChangedModules(changedDirs: List<String>, availableModules: Set<String>): List<String> = changedDirs.flatMap { directory ->
+        val moduleName = directory.replace("/", ":")
+        val ktxModuleName = "$moduleName:ktx"
+        listOf(moduleName, ktxModuleName).filter { availableModules.contains(it) }
+    }
+
+    private fun parseSubProjects(rootDir: File) = File(rootDir, "subprojects.cfg").readLines().filter { !it.startsWith("#") }.toSet()
+
+    @Throws(IOException::class)
+    private fun initRepo(repoPath: File): Git {
+        if (!repoPath.exists()) {
+            throw RuntimeException("Repo path doesn't exist")
+        }
+        return Git.open(repoPath)
+    }
+
+    @Throws(GitAPIException::class)
+    private fun getObjectRefForBranchName(repo: Git, branchName: String): ObjectId {
+        return repo.branchList().setListMode(ListBranchCommand.ListMode.REMOTE).call()
+                .filter { ref: Ref -> ref.name == "refs/remotes/origin/$branchName" }
+                .map { obj: Ref -> obj.objectId }.firstOrNull()
+                ?: throw RuntimeException("Could not find branch named $branchName")
+    }
+
+    private fun getChangedLibraries(repo: Git, startRef: ObjectId, endRef: ObjectId, directories: List<String>): List<String> {
```

Comment:
FWIW, we should probably reduce this to only care about the `CHANGELOG.md` file. It should be the job of people or CI tooling during PRs to catch changed kotlin and java files for releasable artifacts. Otherwise, this could throw false positives- as it may target artifacts in the repo that shouldn't be released.

Reply:
see comment above about widening this to just the path, will try to narrow it from here, changelog.md is a possible narrowing target, but its worth discussing.